src/lib/coaler/embedder/SubstructureAnalyzer.cpp

```diff
@@ -0,0 +1,97 @@
+//
+// Created by chris on 12/7/23.
+//
+
+#include "SubstructureAnalyzer.hpp"
+
+#include <GraphMol/Substruct/SubstructMatch.h>
+#include <GraphMol/new_canon.h>
+
+namespace {
+    struct hasDegreeTwo {
+        bool operator()(RDKit::Atom* atom) { return atom->getDegree() == 2; }
+    };
+
+    /*----------------------------------------------------------------------------------------------------------------*/
+
+    bool allElementsEqual(const std::vector<unsigned> vector) {
+        return std::all_of(vector.begin(), vector.end(), [vector](unsigned i) { return i == vector.at(0); });
+    }
+
+    /*----------------------------------------------------------------------------------------------------------------*/
+
+    std::vector<unsigned> makeSteadyRankVector(const std::vector<unsigned> originalRanks) {
+        std::unordered_map<unsigned, unsigned> originalToSteadyRank;
+        unsigned rank = 0;
+        for (unsigned orignalRank : originalRanks) {
+            if (originalToSteadyRank.count(orignalRank) == 0) {
+                originalToSteadyRank.emplace(orignalRank, rank);
+                rank++;
+            }
+        }
+
+        std::vector<unsigned> newRanks;
+        for (unsigned originalRank : originalRanks) {
+            newRanks.push_back(originalToSteadyRank.at(originalRank));
+        }
+        return newRanks;
+    }
+}  // namespace
```

Comment:
Okay so ganz checke ich den Sinn hiervon nicht. Wir haben eine Kanonisierung (originalRanks) in der jedem Atom ein Rank zugeordnet wird, welcher anscheinend (?) auch mehrfach vergeben werden kann bei Symmetrien.
Wir gehen dann jeden Rank aus originalRanks durch und packen den in eine Map zusammen mit einem hochzählenden neuen rank. Aber nur beim ersten Mal wo wir diesen originalRank finden.
Dann gehen wir nochmal die originalRanks (in der Originalreihenfolge) durch und fügen den dafür vorgesehenen neuen rank in unseren return vector.
Das heißt alles was wir gemacht haben ist das aus z.B. 312312 zu 123123 wird. Die einzige Sinn dahinter für mich in unserem Code wäre, wenn die Kanonisierung nicht bei 0 oder 1 beginnen würde und/oder nicht stetig um 1 steigen würde. Ist das hier der Fall? Wenn ja wer zum Teufel hat so eine Kanonisierung geschrieben?

Reply:
letzteres. bekommst teilweise kanonisierungen wie (0,0,0,4,0,0,0,4) statt (1,1,1,2,1,1,1,2)